cmd/manager/main.go

```diff
@@ -105,96 +137,105 @@ func main() {
 	mgr, err := manager.New(cfg, manager.Options{
 		Namespace:          "",
 		MetricsBindAddress: "0",
+		Scheme:             scheme,
 	})
 	if err != nil {
-		logrus.Fatal(err)
-	}
-
-	logrus.Info("Registering Components.")
-	logrus.Info("setting up scheme")
-	// Setup Scheme for all defined resources
-	if err := apis.AddToScheme(mgr.GetScheme()); err != nil {
-		logrus.Fatal(err)
-	}
-	// Add external resource to scheme
-	if err := olmv1alpha1.AddToScheme(mgr.GetScheme()); err != nil {
-		logrus.Fatal(err)
-	}
-	if err := v1beta1.AddToScheme(mgr.GetScheme()); err != nil {
-		logrus.Fatal(err)
-	}
-	// If the config API is available add the config resources to the scheme
-	if configv1.IsAPIAvailable() {
-		if err := apiconfigv1.AddToScheme(mgr.GetScheme()); err != nil {
-			logrus.Fatal(err)
-		}
-	}
-
-	stopCh := signals.Context().Done()
-
-	var statusReporter status.Reporter = &status.NoOpReporter{}
-	if clusterOperatorName != "" {
-		statusReporter, err = status.NewReporter(cfg, mgr, namespace, clusterOperatorName, os.Getenv("RELEASE_VERSION"), stopCh)
-		if err != nil {
-			logrus.Fatal(err)
-		}
-	}
-
-	// Populate the global default OperatorSources definition and config
-	err = defaults.PopulateGlobals()
-	if err != nil {
-		logrus.Fatal(err)
-	}
-
-	// Setup all Controllers
-	if err := controller.AddToManager(mgr, options.ControllerOptions{}); err != nil {
-		logrus.Fatal(err)
+		logger.Fatal(err)
 	}
 
-	logrus.Info("Setting up health checks")
+	logger.Info("setting up health checks")
 	http.HandleFunc("/healthz", func(w http.ResponseWriter, r *http.Request) {
 		w.WriteHeader(http.StatusOK)
 	})
 	go http.ListenAndServe(":8080", nil)
 
-	// Wait until this instance becomes the leader.
-	logrus.Info("Waiting to become leader.")
-	err = leader.Become(context.TODO(), leaderElectionConfigMapName)
-	if err != nil {
-		logrus.Fatal(err, "Failed to retry for leader lock")
-	}
-	logrus.Info("Elected leader.")
+	ctx := signals.Context()
+	stopCh := ctx.Done()
 
-	logrus.Info("Starting the Cmd.")
+	leaderCtx, leaderCancel := context.WithCancel(context.Background())
+	defer leaderCancel()
 
-	// Handle the defaults
-	err = ensureDefaults(cfg, mgr.GetScheme())
-	if err != nil {
-		logrus.Fatal(err)
-	}
+	run := func(ctx context.Context) {
```

Comment:
The `OnStartedLeading` leader election callback function allows us to specify what should be run once the leader election lease has been acquired, so wrapping the controller/manager specific logic into an inline function, and then calling that inlined function in that callback function allows us to almost completely emulate the previous startup/initialization/etc. logic before migrating to a leaser-for-lease implementation.

There are likely cleaner implementations, like moving this to it's own dedicated function, but going down that path lead to a much larger refactoring of the main executable package and didn't want to inflate the number of changes in this PR.

Reply:
One other things to note as the diff is pretty difficult to parse is that health and liveness probe endpoints are setup _before_ the `run` function is called during the `OnStartedLeading` callback function. It looks like we're blindly setting both of those probes to a 200 status anyways so the current behavior should be retained.